File: python/10-decisiones/seleccion_herramientas.py
"""Selección dinámica de herramientas por similitud Jaccard.

Demuestra el mecanismo de tool retrieval sin dependencias externas:
Jaccard sobre word sets reemplaza embeddings para mostrar el bucle
selección → agente → selección.

Requiere: pip install anthropic

Cómo ejecutar:
    make py SCRIPT=python/10-decisiones/seleccion_herramientas.py

Qué esperar:
    El agente tiene acceso a 20 herramientas pero solo ve las 5 mas relevantes
    para la query actual. Muestra el score Jaccard de cada herramienta candidata.

Variables de entorno:
    MODEL — modelo a usar (default: claude-sonnet-4-6)
"""
import os
import json
import re
from dataclasses import dataclass

import anthropic

MODEL = os.environ.get("MODEL", "claude-sonnet-4-6")
JACCARD_THRESHOLD = 0.04   # sin stemming, palabras distintas reducen el score
TOP_K = 3

SYSTEM_AGENTE = """\
Eres un agente de asistencia. Tienes acceso a un subconjunto de herramientas relevantes
para la tarea actual. Usa las herramientas disponibles para responder.
Si no tienes suficiente información después de una ronda, indica qué necesitarías."""


@dataclass
class Tool:
    name: str
    description: str
    # Texto completo indexado (nombre + descripción + casos de uso)
    index_text: str


# --- Indexación offline ---

def indexar_tools(tools: list[Tool]) -> dict[str, set[str]]:
    """Construye word sets para cada tool. Se ejecuta una sola vez al arrancar."""
    return {
        tool.name: set(tool.index_text.lower().split())
        for tool in tools
    }


# --- Selección en runtime ---

def seleccionar_tools(
    query: str,
    index: dict[str, set[str]],
    tools_by_name: dict[str, Tool],
    k: int = TOP_K,
    threshold: float = JACCARD_THRESHOLD,
) -> list[Tool]:
    query_words = set(query.lower().split())
    scores: list[tuple[str, float]] = []

    for tool_name, tool_words in index.items():
        union = query_words | tool_words
        inter = query_words & tool_words
        score = len(inter) / len(union) if union else 0.0
        scores.append((tool_name, score))

    scores.sort(key=lambda x: x[1], reverse=True)
    return [
        tools_by_name[name]
        for name, score in scores[:k]
        if score >= threshold
    ]


def construir_query_seleccion(tarea: str, ultimo_resultado: str | None) -> str:
    """Combina tarea original con contexto del último resultado para guiar la selección."""
    if not ultimo_resultado:
        return tarea
    # El último resultado orienta qué herramienta se necesita a continuación
    return f"{tarea} — contexto: {ultimo_resultado[:200]}"


# --- Agente con selección dinámica ---

def agente(tarea: str, tools: list[Tool], client: anthropic.Anthropic) -> None:
    index = indexar_tools(tools)
    tools_by_name = {t.name: t for t in tools}
    ultimo_resultado: str | None = None

    print(f"\nTarea: {tarea}")
    print("=" * 60)

    for turno in range(1, 3):
        query_sel = construir_query_seleccion(tarea, ultimo_resultado)
        tools_seleccionadas = seleccionar_tools(query_sel, index, tools_by_name)

        nombres_sel = [t.name for t in tools_seleccionadas]
        print(f"\n[Turno {turno}] Query de selección: {query_sel[:80]}")
        print(f"[Turno {turno}] Tools seleccionadas: {nombres_sel}")

        # Definiciones en formato que Claude entiende
        tool_defs = [
            {
                "name": t.name,
                "description": t.description,
                "input_schema": {
                    "type": "object",
                    "properties": {"query": {"type": "string", "description": "Parámetro de consulta"}},
                    "required": ["query"],
                },
            }
            for t in tools_seleccionadas
        ]

        response = client.messages.create(
            model=MODEL,
            max_tokens=512,
            system=SYSTEM_AGENTE,
            tools=tool_defs,
            messages=[{"role": "user", "content": tarea}],
        )

        # Recopilar lo que el modelo decidió hacer (sin ejecutar realmente las tools)
        acciones = []
        for block in response.content:
            if block.type == "tool_use":
                acciones.append(f"{block.name}({json.dumps(block.input)})")

        if acciones:
            ultimo_resultado = f"Llamadas planeadas: {'; '.join(acciones)}"
            print(f"[Turno {turno}] Acciones: {ultimo_resultado}")
        else:
            texto = next(
                (b.text for b in response.content if hasattr(b, "text")), ""
            )
            ultimo_resultado = texto[:200]
            print(f"[Turno {turno}] Respuesta directa: {ultimo_resultado}")
            break


# --- Catálogo de herramientas ---

TOOLS = [
    Tool(
        name="buscar_contratos",
        description="Busca contratos por nombre de cliente, fecha o estado",
        index_text="buscar contratos cliente acuerdo documento legal renovación fecha vencimiento",
    ),
    Tool(
        name="calcular_fechas",
        description="Calcula diferencias entre fechas, días hasta vencimiento, rangos",
        index_text="calcular fechas diferencia días semanas meses vencimiento plazo duración",
    ),
    Tool(
        name="consultar_crm",
        description="Obtiene información de clientes del CRM: contactos, historial, estado",
        index_text="crm cliente contacto historial estado cuenta empresa organización",
    ),
    Tool(
        name="generar_factura",
        description="Crea facturas con detalle de servicios, impuestos y datos de pago",
        index_text="factura generar crear cobro pago servicio impuesto importe total",
    ),
    Tool(
        name="enviar_email",
        description="Envía correos electrónicos a clientes o equipos internos",
        index_text="email correo enviar notificación mensaje destinatario asunto adjunto",
    ),
    Tool(
        name="analizar_logs",
        description="Analiza logs de sistema para diagnosticar errores y anomalías",
        index_text="logs errores sistema diagnosticar anomalía stack trace excepción servidor",
    ),
]


def main() -> None:
    client = anthropic.Anthropic()

    agente(
        tarea="Busca el contrato de Acme Corp y calcula cuántos días faltan para su renovación",
        tools=TOOLS,
        client=client,
    )


if __name__ == "__main__":
    main()

File: typescript/10-decisiones/seleccion_herramientas.ts
/**
 * Selección dinámica de herramientas por similitud Jaccard.
 *
 * Demuestra el mecanismo de tool retrieval sin dependencias externas:
 * Jaccard sobre word sets reemplaza embeddings para mostrar el bucle
 * selección → agente → selección.
 *
 * Requiere: npm install @anthropic-ai/sdk
 */
import Anthropic from "@anthropic-ai/sdk";

const MODEL = process.env["MODEL"] ?? "claude-sonnet-4-6";
const JACCARD_THRESHOLD = 0.05;
const TOP_K = 3;

const SYSTEM_AGENTE =
  "Eres un agente de asistencia. Tienes acceso a un subconjunto de herramientas " +
  "relevantes para la tarea actual. Usa las herramientas disponibles para responder. " +
  "Si no tienes suficiente información después de una ronda, indica qué necesitarías.";

interface Tool {
  name: string;
  description: string;
  indexText: string;
}

// --- Indexación offline ---

// Cómo ejecutar: make ts SCRIPT=typescript/10-decisiones/seleccion_herramientas.ts


function indexarTools(tools: Tool[]): Map<string, Set<string>> {
  const index = new Map<string, Set<string>>();
  for (const tool of tools) {
    index.set(tool.name, new Set(tool.indexText.toLowerCase().split(/\s+/)));
  }
  return index;
}

// --- Selección en runtime ---

function seleccionarTools(
  query: string,
  index: Map<string, Set<string>>,
  toolsByName: Map<string, Tool>,
  k = TOP_K,
  threshold = JACCARD_THRESHOLD
): Tool[] {
  const queryWords = new Set(query.toLowerCase().split(/\s+/));
  const scores: [string, number][] = [];

  for (const [toolName, toolWords] of index) {
    const union = new Set([...queryWords, ...toolWords]);
    const inter = [...queryWords].filter((w) => toolWords.has(w));
    const score = union.size > 0 ? inter.length / union.size : 0;
    scores.push([toolName, score]);
  }

  scores.sort((a, b) => b[1] - a[1]);

  return scores
    .slice(0, k)
    .filter(([, score]) => score >= threshold)
    .map(([name]) => toolsByName.get(name)!)
    .filter(Boolean);
}

function construirQuerySeleccion(
  tarea: string,
  ultimoResultado: string | null
): string {
  if (!ultimoResultado) return tarea;
  return `${tarea} — contexto: ${ultimoResultado.slice(0, 200)}`;
}

// --- Agente con selección dinámica ---

async function agente(
  tarea: string,
  tools: Tool[],
  client: Anthropic
): Promise<void> {
  const index = indexarTools(tools);
  const toolsByName = new Map(tools.map((t) => [t.name, t]));
  let ultimoResultado: string | null = null;

  console.log(`\nTarea: ${tarea}`);
  console.log("=".repeat(60));

  for (let turno = 1; turno <= 2; turno++) {
    const querySel = construirQuerySeleccion(tarea, ultimoResultado);
    const toolsSel = seleccionarTools(querySel, index, toolsByName);

    console.log(`\n[Turno ${turno}] Query de selección: ${querySel.slice(0, 80)}`);
    console.log(`[Turno ${turno}] Tools seleccionadas: ${toolsSel.map((t) => t.name).join(", ")}`);

    const toolDefs = toolsSel.map((t) => ({
      name: t.name,
      description: t.description,
      input_schema: {
        type: "object" as const,
        properties: {
          query: { type: "string", description: "Parámetro de consulta" },
        },
        required: ["query"],
      },
    }));

    const response = await client.messages.create({
      model: MODEL,
      max_tokens: 512,
      system: SYSTEM_AGENTE,
      tools: toolDefs,
      messages: [{ role: "user", content: tarea }],
    });

    const acciones: string[] = [];
    for (const block of response.content) {
      if (block.type === "tool_use") {
        acciones.push(`${block.name}(${JSON.stringify(block.input)})`);
      }
    }

    if (acciones.length > 0) {
      ultimoResultado = `Llamadas planeadas: ${acciones.join("; ")}`;
      console.log(`[Turno ${turno}] Acciones: ${ultimoResultado}`);
    } else {
      const texto = response.content
        .filter((b) => b.type === "text")
        .map((b) => (b as { type: "text"; text: string }).text)
        .join("");
      ultimoResultado = texto.slice(0, 200);
      console.log(`[Turno ${turno}] Respuesta directa: ${ultimoResultado}`);
      break;
    }
  }
}

// --- Catálogo de herramientas ---

const TOOLS: Tool[] = [
  {
    name: "buscar_contratos",
    description: "Busca contratos por nombre de cliente, fecha o estado",
    indexText: "buscar contratos cliente acuerdo documento legal renovación fecha vencimiento",
  },
  {
    name: "calcular_fechas",
    description: "Calcula diferencias entre fechas, días hasta vencimiento, rangos",
    indexText: "calcular fechas diferencia días semanas meses vencimiento plazo duración",
  },
  {
    name: "consultar_crm",
    description: "Obtiene información de clientes del CRM: contactos, historial, estado",
    indexText: "crm cliente contacto historial estado cuenta empresa organización",
  },
  {
    name: "generar_factura",
    description: "Crea facturas con detalle de servicios, impuestos y datos de pago",
    indexText: "factura generar crear cobro pago servicio impuesto importe total",
  },
  {
    name: "enviar_email",
    description: "Envía correos electrónicos a clientes o equipos internos",
    indexText: "email correo enviar notificación mensaje destinatario asunto adjunto",
  },
  {
    name: "analizar_logs",
    description: "Analiza logs de sistema para diagnosticar errores y anomalías",
    indexText: "logs errores sistema diagnosticar anomalía stack trace excepción servidor",
  },
];

async function main(): Promise<void> {
  const client = new Anthropic();

  await agente(
    "Busca el contrato de Acme Corp y calcula cuántos días faltan para su renovación",
    TOOLS,
    client
  );
}

main().catch(console.error);
